pkg/alertmanager/api_grafana.go

```diff
@@ -168,6 +171,60 @@ type errorResult struct {
 	Error  string `json:"error"`
 }
 
+func (am *MultitenantAlertmanager) GetUserState(w http.ResponseWriter, r *http.Request) {
+	logger := util_log.WithContext(r.Context(), am.logger)
+
+	userID, err := tenant.TenantID(r.Context())
+	if err != nil {
+		level.Error(logger).Log("msg", errNoOrgID, "err", err.Error())
+		w.WriteHeader(http.StatusUnauthorized)
+		util.WriteJSONResponse(w, errorResult{Status: statusError, Error: fmt.Sprintf("%s: %s", errNoOrgID, err.Error())})
+		return
+	}
+
+	// Try to get the full state from the running Alertmanager, using the database as a fallback.
+	am.alertmanagersMtx.Lock()
+	userAM, ok := am.alertmanagers[userID]
+	am.alertmanagersMtx.Unlock()
```

Comment:
I wonder if safer approach would be to keep the lock until we're done reading the state because once we release it there is a chance the Alertmanager gets stopped, which might do some mangling with the state.